configuration: add logging section & ad log_level to configuration_opts

olaf-mandel opened the pull request:
Allow to configure both the logging-levels and targets of various loggers by adding a new section `logging` to the configuration file. Each entry in that section configures one logger / channel. An example config is:
    
```yaml
logging:
  - name root
    level: INFO
  - name: foomodule.barcollector:
    level: WARNING
    target: /path/to/my/collector/logfile.log
```

Add `CollectorBase.setLoggers(logger_names)` as a means for collectors derived from `CollectorBase` to easily configure logging-levels. The new method uses the `collector_opts.logging_level` value (if set in the configuration) to set the logging-level of any loggers given to it.

cmeissner commented on README.md:
```diff
@@ -113,6 +113,9 @@ collector_opts:
       - lo
+logging_levels:
+  root: INFO
+  foomodule.barcollector: WARNING
```
This is not that optimal. Configuration for should be placed in `collector_opts.<collector_name>`. That leads to a structure sth like this:

```
collector_opts:
  example:
    log_level: WARNING
``` 

Can you please take this change in consideration?

olaf-mandel replied:
I agree that configuring the logging levels in the `collector_opts` for each collector would be semantically cleaner. But that would require changes to all collector-implementations out there: because there is no correlation between the name of a logger / channel and the  name of the python-package that uses it, this couldn't be implemented by `p3exporter`. In fact, in my use-case I am configuring not the logging of the actual collector but of a package used by that collector...

The requested change would require all classes derived from `CollectorBase` to parse this extra entry in the `CollectorConfig.collector_opts`. And that is assuming that there is no collector out there that already implemented this setting but with a different semantic, e.g. as a dict of sub-modules...

I am open to suggestions how to make the requested change locally in `p3exporter` if I am overlooking something.

cmeissner replied:
The change would only affect the `CollectorBase` class and such, it can provide methods for managing logging in each collector.

olaf-mandel replied:
You mean like adding a method `CollectorBase.setLoggers(logger_names)` that the deriving collectors call from their constructors? Maybe add an optional argument to `CollectorBase.__init__()` as a shortcut? I can do that (please indicate if I understood you correctly and I will start on that).

But as this depends on the collectors being updated to use this. I would therefore like to keep the global `logging_levels` config in addition to the above per-collector one.

cmeissner replied:
> please indicate if I understood you correctly and I will start on that

`CollectorBase.setLoggers(logger_names)` is what I thought of. Thank you for implementing. It would be great if you also show how to apply the new method in the distributed collectors.

> I would therefore like to keep the global `logging_levels` config in addition

Okay, but I would have some adjustments to be more flexible. Starting from the configuration, I would prefer a structure like this:

```
logging:
  - name: <class>:
    level: <LEVEL>
    target: <target file, default to STDOUT>
```

With this, the structure of your proposal would look as followed:

```
logging:
  - name root
    level: INFO
  - name: foomodule.barcollector: 
    level: WARNING
    target: /path/to/my/collector/logfile.log
```

Can you deal with that extension to your idea?

olaf-mandel replied:
Will take a bit if time, but will do so.

olaf-mandel replied:
Done.

cmeissner replied:
Thank you for your effort. Implementing the requested changes in under 6 hours is really a good pace. 

May I also ask for your opinion? Do you mean that `logging[].name` is a good choice, or should it better named `logging[].class_name`?

olaf-mandel replied:
I think `name` is a good choice. And specifically compared to `class_name`: these are arbitrary names given to loggers, I could call one "Egg, bacon, sausage, and Spam" if I really wanted to. The only special character in there is the dot (`.`) which is used to determine the hierarchy of loggers, AFAICT.

The only contenders from my POV would be `logger_name` and `channel_name`, but I think `name` is still preferable here.

cmeissner commented on p3exporter/__init__.py:
```diff
@@ -41,6 +41,13 @@ def main():
     with open(args.config, 'r') as config_file:
         cfg = yaml.load(config_file, Loader=yaml.SafeLoader)
     collector_config = CollectorConfig(**cfg)
+    if type(cfg.get('logging_levels')) == dict:
```
Regarding to codacy results using `is_instance` should be prefered over `type`

olaf-mandel replied:
Changed line to `if isinstance(cfg.get('logging_levels'), dict):` now.

cmeissner commented:
As there is new code introduced, please also provide tests to check the funtionallity.

cmeissner reviewed (changes requested):
Please adapt as proposed

olaf-mandel commented:
Refactored for testability and added test.

olaf-mandel committed:
configuration: add logging section

Allow to configure both the logging-levels and targets of various
loggers by adding a new section `logging` to the configuration file.
Each entry in that section configures one logger / channel. An example
config is:

```yaml
logging:
  - name root
    level: INFO
  - name: foomodule.barcollector:
    level: WARNING
    target: /path/to/my/collector/logfile.log
```

olaf-mandel committed:
configuration: add log_level to configuration_opts

Add `CollectorBase.setLoggers(logger_names)` as a means for collectors
derived from `CollectorBase` to easily configure logging-levels. The new
method uses the `collector_opts.logging_level` value (if set in the
configuration) to set the logging-level of any loggers given to it.

cmeissner commented on p3exporter/__init__.py:
```diff
@@ -13,6 +13,35 @@
+            return
+        logger = logging.getLogger(c["name"])
+        level = c.get('level')
+        if not level is None:
```
According to flake8 here, `is not` should be preferred over `not ... is`. It will increase the readability of the code.

olaf-mandel replied:
Done.

olaf-mandel replied:
Done.

cmeissner commented on p3exporter/__init__.py:
```diff
@@ -13,6 +13,35 @@
+        if not level is None:
+            logger.setLevel(level)
+        target = c.get('target')
+        if not target is None:
```
According to flake8 here, `is not` should be preferred over `not ... is`. It will increase the readability of the code.

olaf-mandel replied:
Done.

cmeissner commented on p3exporter/collector/__init__.py:
```diff
@@ -58,6 +58,25 @@ def collector_name_from_class(self):
         return '_'.join(class_name_parts)
 
+    def setLoggers(self, logger_names: list | str):
+        """Configure the provided logger(s) according to the CollectorConfig
```
According to flake8, `First line should end with a periodFlake8(D400)`.

olaf-mandel replied:
Done.

cmeissner reviewed (changes requested):
There are some `flake8` errors and warnings in your PR. It would be great if you can fix this too.

cmeissner committed:
Merge branch 'develop' into feature/config-logging_levels
